tracer/src/Datadog.Trace/Activity/Handlers/DefaultActivityHandler.cs

```diff
@@ -0,0 +1,291 @@
+                            // for propagation then the current active span will appear as parentId
+                            // in the context propagation, and we will keep the entire trace.
+                            // TraceId
+                            if (string.IsNullOrWhiteSpace(activeSpan.Context.RawTraceId))
+                            {
+                                w3cAct.TraceId = activeSpan.TraceId.ToString("x32");
+                            }
+                            else
+                            {
+                                w3cAct.TraceId = activeSpan.Context.RawTraceId;
+                            }
+
+                            // SpanId
+                            if (string.IsNullOrWhiteSpace(activeSpan.Context.RawSpanId))
+                            {
+                                w3cAct.ParentSpanId = activeSpan.SpanId.ToString("x16");
+                            }
+                            else
+                            {
+                                w3cAct.ParentSpanId = activeSpan.Context.RawSpanId;
+                            }
+
+                            // We clear internals Id and ParentId values to force recalculation.
+                            w3cAct.RawId = null;
+                            w3cAct.RawParentId = null;
+                        }
+
+                        return;
+                    }
+                }
+
+                lock (ActivityScope)
+                {
+                    if (!ActivityScope.TryGetValue(activity.Instance, out _))
+                    {
+                        var span = Tracer.Instance.StartSpan(activity.OperationName, startTime: activity.StartTimeUtc, traceId: traceId, spanId: spanId, rawTraceId: rawTraceId, rawSpanId: rawSpanId);
+                        var scope = Tracer.Instance.ActivateSpan(span, false);
+                        ActivityScope[activity.Instance] = scope;
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error processing the OnActivityStarted callback");
+            }
+        }
+
+        public void ActivityStopped<T>(string sourceName, T activity)
+            where T : IActivity
+        {
+            try
+            {
+                var hasActivity = activity?.Instance is not null;
+                if (hasActivity)
+                {
+                    foreach (var ignoreSourceName in IgnoreOperationNamesStartingWith)
+                    {
+                        if (activity!.OperationName?.StartsWith(ignoreSourceName) == true)
+                        {
+                            return;
+                        }
+                    }
+                }
+
+                lock (ActivityScope)
+                {
+                    if (hasActivity && ActivityScope.TryGetValue(activity!.Instance!, out var scope) && scope?.Span is not null)
```

Comment:
You must hate me for suggesting `nullable enable` - I feel like you're not a convert 😂

Reply:
the analyzer is pretty dumb! :D

Reply:
I think it's got you on a technicality here, `Instance` _could_ be null theoretically if it's been changed in a background thread 😉